osu.Game/Rulesets/Mods/ModAdaptiveSpeed.cs

```diff
@@ -0,0 +1,188 @@
+        private const int average_count = 6;
+
+        public override string Name => "Adaptive Speed";
+
+        public override string Acronym => "AS";
+
+        public override string Description => "Let track speed adapt to you.";
+
+        public override ModType Type => ModType.Fun;
+
+        public override double ScoreMultiplier => 1;
+
+        public override Type[] IncompatibleMods => new[] { typeof(ModRateAdjust), typeof(ModTimeRamp) };
+
+        [SettingSource("Initial rate", "The starting speed of the track")]
+        public BindableNumber<double> InitialRate { get; } = new BindableDouble
+        {
+            MinValue = 0.5,
+            MaxValue = 2,
+            Default = 1,
+            Value = 1,
+            Precision = 0.01
+        };
+
+        [SettingSource("Adjust pitch", "Should pitch be adjusted with speed")]
+        public BindableBool AdjustPitch { get; } = new BindableBool
+        {
+            Default = true,
+            Value = true
+        };
+
+        public BindableNumber<double> SpeedChange { get; } = new BindableDouble
+        {
+            Default = 1,
+            Value = 1
+        };
+
+        private ITrack track;
+        private double targetRate = 1d;
+
+        private readonly List<double> recentRates = Enumerable.Repeat(1d, average_count).ToList();
+
+        // rate for a hit is calculated using the end time of another hit object earlier in time
+        // caching them here for easy access
+        private readonly Dictionary<HitObject, double> previousEndTimes = new Dictionary<HitObject, double>();
+
+        // record the value removed from recentRates when an object is hit
+        // for rewind support
+        private readonly Dictionary<HitObject, double> dequeuedRates = new Dictionary<HitObject, double>();
+
+        public ModAdaptiveSpeed()
+        {
+            InitialRate.BindValueChanged(val =>
+            {
+                SpeedChange.Value = val.NewValue;
+                targetRate = val.NewValue;
+            });
+            AdjustPitch.BindValueChanged(applyPitchAdjustment);
+        }
+
+        public void ApplyToTrack(ITrack track)
+        {
+            this.track = track;
+
+            InitialRate.TriggerChange();
+            AdjustPitch.TriggerChange();
+            recentRates.Clear();
+            recentRates.AddRange(Enumerable.Repeat(InitialRate.Value, average_count));
+        }
+
+        public void ApplyToSample(DrawableSample sample)
+        {
+            sample.AddAdjustment(AdjustableProperty.Frequency, SpeedChange);
+        }
+
+        public void Update(Playfield playfield)
+        {
+            SpeedChange.Value = Interpolation.DampContinuously(SpeedChange.Value, targetRate, 50, playfield.Clock.ElapsedFrameTime);
+        }
+
+        public double ApplyToRate(double time, double rate = 1) => rate * InitialRate.Value;
+
+        private void applyPitchAdjustment(ValueChangedEvent<bool> adjustPitchSetting)
```

Comment:
Rename to `onAdjustPitchChanged` or `adjustPitchChanged`. This is still misleading to my eyes.